hudi-client/src/main/java/org/apache/hudi/index/HoodieSimpleIndex.java

```diff
@@ -0,0 +1,202 @@
+    jsc.setJobGroup(this.getClass().getSimpleName(), "Fetching fileInfos for partitions of interest");
+    List<String> affectedPartitionPathList = partitionRecordKeyPairRDD.map(tuple -> tuple._1).distinct().collect();
+    JavaRDD<Tuple2<String, String>> fileInfoList = jsc.parallelize(
+        loadAllFilesForPartitions(affectedPartitionPathList, jsc, hoodieTable)).sortBy(Tuple2::_1, true, config.getSimpleIndexParallelism());
+
+    JavaPairRDD<String, String> partitionFileIndexInfoPairRDD = fileInfoList.mapToPair(entry -> new Tuple2<>(entry._1, entry._2));
+
+    jsc.setJobGroup(this.getClass().getSimpleName(), "Fetching records from all files of interest");
+    return partitionFileIndexInfoPairRDD.flatMapToPair(
+        (PairFlatMapFunction<Tuple2<String, String>, HoodieKey, HoodieRecordLocation>) partitionPathFileId ->
+            new RecordFetcher(partitionPathFileId, hoodieTable).getResultSet()
+    );
+  }
+
+  /**
+   * Load all involved files as <Partition, filename> pair RDD.
+   */
+  private List<Tuple2<String, String>> loadAllFilesForPartitions(List<String> partitions, final JavaSparkContext jsc,
+                                                                 final HoodieTable hoodieTable) {
+
+    // Obtain the latest data files from all the partitions.
+    List<Pair<String, String>> partitionPathFileIDList = loadLatestDataFilesForAllPartitions(partitions,
+        jsc, hoodieTable);
+    return partitionPathFileIDList.stream()
+        .map(pf -> new Tuple2<>(pf.getKey(), pf.getValue())).collect(toList());
+  }
+
+  /**
+   * Fetch the latest base file for the given partition and fileId.
+   *
+   * @param hoodieTable           instance of {@link HoodieTable} in which the partition exists
+   * @param partitionPathFilePair Partition path fileId pair
+   * @return the latest data file for the given partition and fileId
+   */
+  private HoodieBaseFile getLatestDataFile(HoodieTable hoodieTable, Pair<String, String> partitionPathFilePair) {
+    return hoodieTable.getBaseFileOnlyView()
+        .getLatestBaseFile(partitionPathFilePair.getLeft(), partitionPathFilePair.getRight()).get();
+  }
+
+  /**
+   * Record Fetcher for a given partitionPath, fileId pair.
+   */
+  class RecordFetcher {
+    private HoodieTable<T> table;
+    private Tuple2<String, String> partitionPathFileIdPair;
+
+    RecordFetcher(Tuple2<String, String> partitionPathFileIdPair, HoodieTable<T> table) {
+      this.partitionPathFileIdPair = partitionPathFileIdPair;
+      this.table = table;
+    }
+
+    Iterator<Tuple2<HoodieKey, HoodieRecordLocation>> getResultSet() throws Exception {
+      HoodieBaseFile baseFile = getLatestDataFile(table, Pair.of(partitionPathFileIdPair._1, partitionPathFileIdPair._2));
+      List<Pair<HoodieKey, HoodieRecordLocation>> records = ParquetUtils.fetchRecordKeyPartitionPathFromParquet(table.getHadoopConf(), new Path(baseFile.getPath()),
```

Comment:
we can't call `ParquetUtils` from here.. that breaks abstraction. Lets design a new handle

Reply:
RecordFetcher is the actual handle. Already I have abstracted out. not sure what more do we need. Can you help me clarify @vinothchandar

Reply:
what I mean by handle is a subclass of `HoodieReadHandle` .. it's an anti pattern to assume its a parquet base file from the index level.. Let's fix this